blockchain/chain/src/store/chain_store.rs

```diff
@@ -610,6 +695,87 @@ pub fn draw_randomness(
     Ok(ret)
 }
 
+/// Returns the heaviest tipset
+pub fn get_heaviest_tipset<DB>(db: &DB) -> Result<Option<Tipset>, Error>
+where
+    DB: BlockStore,
+{
+    match db.read(HEAD_KEY)? {
+        Some(bz) => {
+            let keys: Vec<Cid> = from_slice(&bz)?;
+            Ok(Some(tipset_from_keys(db, &TipsetKeys::new(keys))?))
+        }
+        None => Ok(None),
+    }
+}
+
+/// Returns Tipset from key-value store from provided cids
+pub fn tipset_from_keys<DB>(db: &DB, tsk: &TipsetKeys) -> Result<Tipset, Error>
+where
+    DB: BlockStore,
+{
+    if tsk.cids().is_empty() {
+        Ok(get_heaviest_tipset(db)?.ok_or_else(|| Error::NotFound("Heaviest Tipset not found"))?)
+    } else {
+        let block_headers: Vec<BlockHeader> = tsk
+            .cids()
+            .iter()
+            .map(|c| {
+                db.get(c)
+                    .map_err(|e| Error::Other(e.to_string()))?
+                    .ok_or_else(|| Error::NotFound("Key for header"))
+            })
+            .collect::<Result<_, Error>>()?;
+
+        // construct new Tipset to return
+        let ts = Tipset::new(block_headers)?;
+        Ok(ts)
+    }
+}
+/// Returns the tipset behind `tsk` at a given `height`.
+/// If the given height is a null round:
+/// - If `prev` is `true`, the tipset before the null round is returned.
+/// - If `prev` is `false`, the tipset following the null round is returned.
+///
+/// Returns `None` if the tipset provided was the tipset at the given height.
+pub fn tipset_by_height<DB>(
+    db: &DB,
+    height: ChainEpoch,
+    ts: &Tipset,
+    prev: bool,
+) -> Result<Option<Tipset>, Error>
+where
+    DB: BlockStore,
+{
+    if height > ts.epoch() {
+        return Err(Error::Other(
+            "searching for tipset that has a height less than starting point".to_owned(),
+        ));
+    }
+    if height == ts.epoch() {
+        return Ok(Some(ts.clone()));
+    }
+    // TODO: If ts.epoch()-h > Fork Length Threshold, it could be expensive to look up
+    let mut ts_temp: Option<Tipset> = None;
+    loop {
+        let pts = if let Some(temp) = &ts_temp {
+            tipset_from_keys(db, temp.parents())?
+        } else {
+            tipset_from_keys(db, ts.parents())?
+        };
+        if height > pts.epoch() {
+            if prev {
+                return Ok(Some(pts));
+            }
+            return Ok(ts_temp);
+        }
+        if height == pts.epoch() {
+            return Ok(Some(pts));
+        }
+        ts_temp = Some(pts);
+    }
+}
```

Comment:
please revert this as well, shouldn't have a regression in this (there are the functions attached to `ChainStore` you can use